packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,394 @@
+function internalUpgrade(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instancedDefinition: Definition
+) {
+    setPrototypeOf(instance, instancedDefinition.UserCtor.prototype);
+    definitionForElement.set(instance, instancedDefinition);
+    // attributes patches when needed
+    if (instancedDefinition !== registeredDefinition) {
+        patchAttributes(instance, registeredDefinition, instancedDefinition);
+    }
+    // Tricking the construction path to believe that a new instance is being created,
+    // that way it will execute the super initialization mechanism but the HTMLElement
+    // constructor will reuse the instance by returning the upgradingInstance.
+    // This is by far the most important piece of the puzzle
+    upgradingInstance = instance;
+    try {
+        new instancedDefinition.UserCtor();
+    } catch (err) {
+        if (err instanceof TypeError && err.message === 'Illegal constructor') {
+            new instancedDefinition.UserCtor();
+        } else {
+            throw err;
+        }
+    }
+
+    const { observedAttributes, attributeChangedCallback } = instancedDefinition;
+    if (observedAttributes.size === 0 || isUndefined(attributeChangedCallback)) {
+        return;
+    }
+    const offset = getObservedAttributesOffset(registeredDefinition, instancedDefinition);
+    if (offset.size === 0) {
+        return;
+    }
+    // Approximate observedAttributes from the user class, but only for the offset attributes
+    offset.forEach((name) => {
+        if (nativeHasAttribute.call(instance, name)) {
+            const newValue = nativeGetAttribute.call(instance, name);
+            attributeChangedCallback!.call(instance, name, null, newValue);
+        }
+    });
+}
+
+function getDefinitionForConstructor(constructor: CustomElementConstructor): Definition {
+    if (!constructor || !constructor.prototype || typeof constructor.prototype !== 'object') {
+        throw new TypeError(`The referenced constructor is not a constructor.`);
+    }
+    let definition = definitionForConstructor.get(constructor);
+    if (isUndefined(definition)) {
+        definition = createDefinitionRecord(constructor);
+        definitionForConstructor.set(constructor, definition);
+    }
+    return definition;
+}
+
+export function patchCustomElementRegistry() {
+    const { customElements: nativeRegistry } = window;
+    const { define: nativeDefine, whenDefined: nativeWhenDefined, get: nativeGet } = nativeRegistry;
+
+    if ((nativeRegistry as any)[registryPatchedSymbol]) {
+        throw new Error(
+            'patchCustomElementRegistry was called twice in the same JavaScript environment. ' +
+                'Please ensure you are loading the LWC engine only once. ' +
+                'If this is a Jest environment, check your Jest/JSDOM configuration.'
+        );
+    }
+    (nativeRegistry as any)[registryPatchedSymbol] = true;
```

Comment:
If we don't add this check, then an error will be thrown eventually for "Illegal constructor." I saw this in two of our downstreams' Jest tests.

I considered wrapping this in a `process.env.NODE_ENV` condition, but what if a consumer loads two versions of the engine and one is minified? Better safe than sorry.

Reply:
this is one of the things that I want to discuss, this might actually be a red flag.

Reply:
Chatted with Caridy, the concerns are:

1. If the patched registry doesn't work when applied twice, then this may limit our ability to have multiple versions of the framework running on top of the same registry. Ideally it should be resilient to being applied twice.
2. #2799 might allow us to get rid of the `UserCtor` being passed in to the `PivotCtor`, we should explore that as well:


https://github.com/salesforce/lwc/blob/5541855baba494d1f3c54be35eda3aa8e969e856/packages/%40lwc/engine-dom/src/patches/global-registry.ts#L59-L60

Reply:
Good news, it works with two copies of the engine when running in Karma. Looks like this is a Jest-only issue.